`'static` lifetime on hook init fns are too restrictive

**Problem**

Hooks with initialization functions have a 'static lifetime requirement
on the signature, however, any closure given only needs to have a
lifetime that lives as long as the hook call.

This issue applies to the following hooks:
- `use_state`
- `use_reducer_with_init`
- `use_ref`

**Steps To Reproduce**

The `'static` lifetime makes this code:

```rust
#[function_component(Comp)]
fn comp(props: &Props) -> Html {
  let state = use_state(|| props.value.clone());
                           ^^^^^^^^^^^^
  //.. then continue
}
```

fail to compile with the following error:

```bash
error[E0759]: `props` has an anonymous lifetime `'_` but it needs to satisfy a `'static` lifetime requirement
  --> src/main.rs:14:23
   |
13 | fn comp(props: &Props) -> Html {
   |                ------ this data with an anonymous lifetime `'_`...
14 |     let _ = use_state(|| props.value.clone());
   |                       ^^^^^^^^^^^^^^^^^^^^^^ ...is captured here...
   |
note: ...and is required to live as long as `'static` here
  --> src/main.rs:14:13
   |
14 |     let _ = use_state(|| props.value.clone());
   |             ^^^^^^^^^
```

_Example above posed as [a question in Discord](https://discord.com/channels/701068342760570933/703449306497024049/884325662305964082)_ 

`&Props` doesn't have a `'static` lifetime but an anonymous one which is tied
to the scope of the function (`comp` in this case), therefore the reference 
captured by the closure doesn't satisfy the `'static` requirement. 

**Expected behavior**
For initializing functions of hooks to have an anonymous lifetime so that
prop values can be used to initialize state, ref, and reducer.

**Environment:**
 - Yew version:  `master`

**Questionnaire**

- I would like to fix and I have a solution

--- a/packages/yew/src/functional/hooks/use_reducer.rs
+++ b/packages/yew/src/functional/hooks/use_reducer.rs
@@ -118,7 +118,7 @@ where
     Action: 'static,
     State: 'static,
     InitialState: 'static,
-    InitFn: Fn(InitialState) -> State + 'static,
+    InitFn: Fn(InitialState) -> State,
 {
     let init = Box::new(init);
     let reducer = Rc::new(reducer);

--- a/packages/yew/src/functional/hooks/use_ref.rs
+++ b/packages/yew/src/functional/hooks/use_ref.rs
@@ -46,7 +46,7 @@ use std::{cell::RefCell, rc::Rc};
 ///     }
 /// }
 /// ```
-pub fn use_ref<T: 'static>(initial_value: impl FnOnce() -> T + 'static) -> Rc<RefCell<T>> {
+pub fn use_ref<T: 'static>(initial_value: impl FnOnce() -> T) -> Rc<RefCell<T>> {
     use_hook(
         || Rc::new(RefCell::new(initial_value())),
         |state, _| state.clone(),

--- a/packages/yew/src/functional/hooks/use_state.rs
+++ b/packages/yew/src/functional/hooks/use_state.rs
@@ -34,7 +34,7 @@ struct UseState<T2> {
 ///     }
 /// }
 /// ```
-pub fn use_state<T: 'static, F: FnOnce() -> T + 'static>(initial_state_fn: F) -> UseStateHandle<T> {
+pub fn use_state<T: 'static, F: FnOnce() -> T>(initial_state_fn: F) -> UseStateHandle<T> {
     use_hook(
         // Initializer
         move || UseState {

